CONTRIBUTING.md

```diff
@@ -1,62 +1,55 @@
-- Write codes for new features or bug fixes.
+If there isn't an issue opening for what you are going to work on, please open an issue before writing codes for a pull request, for these reasons:
 
-## Your First Contribution
+1.  You may get help from others.
+2.  Others can know you are working on it, this will save possible duplicate works.
+3.  You will be notified if this feature is not going to be accepted, or the "bug" is actually not a bug, etc., thus saving your time.
 
-**Working on your first Pull Request?** You can learn how from this *free* series [How to Contribute to an Open Source Project on GitHub](https://egghead.io/series/how-to-contribute-to-an-open-source-project-on-github). Basically, you have to fork this repository, write codes on your fork, open a pull request when you are finished, wait for reviews and apply suggestions from the maintainers.
+## Contributing rules
 
-**First time to developing?** There's a big difference between developing and competitive programming, but it's not hard. The first thing you should know is Object-Oriented Programming. Basically, you need to know how classes work in C++, and you should know it's a rule to put the class definitions in the `.h`/`.hpp` files, and put the implementation of the functions in the `.cpp` files. The source code of the CP Editor is a good example. Google for what you want to implement is usually helpful. Copy and modify the source codes if you want to implement something similar to an existing feature. If you are looking for an IDE for developing, [VS Code](https://code.visualstudio.com/) is probably a good choice.
+-   Remember that the goal is to build an editor for competitive programmers, not for developers.
+-   Do one thing in a pull request. If you want to add multiple features/fix multiple bugs and they are not relevant, open multiple pull requests. You may need to create multiple branches to open multiple pull requests, you can learn git branching [here](https://learngitbranching.js.org/).
+-   Add new features on the master branch, and fix bugs on the latest active (master, beta and stable) branch which has the bug. See also [the versioning rule](VERSIONING.md).
+-   Please write [clear commit messages](https://chris.beams.io/posts/git-commit/). It's also recommended to write [Conventional Commits](https://www.conventionalcommits.org/). Since we are [squashing and merging pull requests](https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/about-pull-request-merges#squash-and-merge-your-pull-request-commits), the commit messages in a pull request is not so important, while the title of the pull request or the message of the only commit in a pull request is more crucial.
+-   Use Clang Format and [`.clang-format`](.clang-format) to format your codes. If the codes without your modifications are changed after formatting, try updating your `clang-format`.
```

Comment:
But I'm always using the latest version on Arch. So if there are any breaking changes, we'll update the CI instead of using the same version as CI.

Reply:
> But I'm always using the latest version on Arch. So if there are any breaking changes, we'll update the CI instead of using the same version as CI.

Okay, So I have changed the statement to "same or higher version of clang-format than used in CI"